### PR TITLE
Some improvements

### PR DESCRIPTION
**Pull request by metaworm**
* Codegen for `break` statement
* Debug information parsing, resolve line number and function name when error raises

### CODE REVIEW

File: src/binary_chunk/deserialize.rs [hunk 1]
```diff
@@ -96,17 +96,24 @@ fn load_function<'gc, R: Read>(
 
     // AbsLineInfo
     let n = load_int(reader)?;
+    let mut abslineinfo = Vec::with_capacity(n as _);
```
Comment (mosmeh) [inline src/binary_chunk/deserialize.rs]: Please use Rust style variable / function name, such as abs_line_info
Same for other parts. Do not just copy and paste names from the original Lua code.

File: src/codegen.rs [hunk 2]
```diff
@@ -39,6 +39,12 @@ pub enum CodegenError {
     #[error("cannot use '...' outside a vararg function")]
     VarArgExpressionOutsideVarArgFunction,
 
+    #[error("not in a loop block")]
+    NotInLoop,
+
+    #[error("mismatched block")]
+    MismatchedBlock,
+
```
Comment (mosmeh) [inline src/codegen.rs]: Please use same error messages as the original Lua.
For `NotInLoop`, its "break outside loop". So `BreakOutsideLoop`?

File: src/codegen.rs [hunk 3]
```diff
@@ -241,10 +247,17 @@ impl Frame<'_> {
     }
 }
 
+#[derive(Default)]
+struct LoopInfo {
+    pub break_label: Option<Label>,
+    pub continue_label: Option<Label>,
```
Comment (mosmeh) [inline src/codegen.rs]: They don't have to be `pub`.
Also `continus_label` is unused.

File: src/codegen.rs [hunk 4]
```diff
@@ -241,10 +247,17 @@ impl Frame<'_> {
     }
 }
 
+#[derive(Default)]
+struct LoopInfo {
```
Comment (mosmeh) [inline src/codegen.rs:251]: Can't it be generic scopes / blocks rather than being specific to loops?

File: src/codegen/ast.rs [hunk 5]
```diff
@@ -218,7 +218,10 @@ impl<'gc> CodeGenerator<'gc> {
             Statement::LocalFunction(s) => self.codegen_local_func_statement(s)?,
             Statement::LocalVariable(s) => self.codegen_local_variable_statement(s)?,
             Statement::Label(_) => todo!("label"),
-            Statement::Break => todo!("break"),
+            Statement::Break => {
+                let label = self.break_label()?;
+                self.emit(IrInstruction::Jump { target: label });
+            }
```
Comment (mosmeh) [inline src/codegen/ast.rs]: Please follow the style of other statements: `codegen_*_statement`

File: src/runtime.rs [hunk 6]
```diff
@@ -321,6 +330,17 @@ impl<'gc> LuaThread<'gc> {
             _ => unreachable!(),
         }
     }
+
+    fn last_lua_frame(&self) -> Option<&LuaFrame> {
```
Comment (mosmeh) [inline src/runtime.rs]: current rather than last?

File: src/runtime.rs [hunk 7]
```diff
@@ -321,6 +330,17 @@ impl<'gc> LuaThread<'gc> {
...
+        match self.frames.as_slice() {
+            [.., Frame::Lua(frame)] => Some(frame),
+            _ => None,
+        }
+    }
+
+    fn stack_closure(&self, n: usize) -> Option<&'_ LuaClosure<'gc>> {
```
Comment (mosmeh) [inline src/runtime.rs:345]: This function is only used in single place and doesn't look generally useful

File: src/runtime/debug.rs [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        // }
+        let frame = thread.last_lua_frame()?;
+        let closure = thread.stack_closure(frame.bottom)?;
+        closure.proto.funcname_from_code(frame.last_pc())
+    }
+}
+
+impl<'gc> LuaClosureProto<'gc> {
+    pub(crate) fn funcname_from_code(&self, pc: usize) -> Option<(&'static str, &'_ str)> {
+        let i = self.code.get(pc)?;
```
Comment (mosmeh) [inline src/runtime/debug.rs]: `i` sounds like a loop variable. `insn` is preferred

File: src/runtime/debug.rs [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    opcode::{self, OpCode},
+    Instruction, LuaFrame, Metamethod, Vm,
+};
+
+impl<'gc> Vm<'gc> {
+    pub(crate) fn funcname_from_call<'a>(
+        &self,
+        thread: &'a mut LuaThread<'gc>,
+        bottom: usize,
+    ) -> Option<(&'static str, &'a str)> {
```
Comment (mosmeh) [inline src/runtime/debug.rs]: It's not clear what the return value means. How about introducing a struct?

File: src/runtime/debug.rs [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    fn gxf(&self, pc: usize, i: Instruction, isup: bool) -> &'static str {
+        let t = i.b(); /* table index */
+        // const char *name;  /* name of indexed variable */
+        let name = if isup {
+            /* is an upvalue? */
+            self.upvalname(t)
+        } else {
+            self.get_objname(pc, t).map(|x| x.1)
+        };
+        if name.filter(|&n| n == "_ENV").is_some() {
```
Comment (mosmeh) [inline src/runtime/debug.rs]: Can be replaced with pattern match?

File: src/runtime/opcode.rs [hunk 11]
```diff
@@ -119,3 +119,130 @@ opcodes! {
...
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct Modes {
+    pub mode: OpMode,
+    pub mm: bool,
+    pub ot: bool,
+    pub it: bool,
+    pub test: bool,
+    pub set_a: bool,
```
Comment (mosmeh) [inline src/runtime/opcode.rs]: These names are too cryptic

File: src/types/function.rs [hunk 12]
```diff
@@ -41,6 +45,18 @@ impl NativeFunction {
...
 }
 
+#[derive(Debug, Clone, Copy)]
+pub struct AbsLineInfo {
+    pub pc: u32,
+    pub line: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct LocalVar<'gc> {
```
Comment (mosmeh) [inline src/types/function.rs]: Please use `LocalVariable` (if you want to make it Rusty) or `LocVar` (if you keep it C-ish)

File: src/runtime/debug.rs [hunk 13]
```diff
@@ -0,0 +1,295 @@
+use crate::types::{AbsLineInfo, LuaClosureProto, LuaThread, Value};
+
+use super::{
+    opcode::{self, OpCode},
+    Instruction, LuaFrame, Metamethod, Vm,
+};
+
```
Comment (mosmeh) [inline src/runtime/debug.rs:7]: Some of the functions don't have to be `pub`

File: src/runtime/debug.rs [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+            //     let b = if op == OpCode::LOADK {
+            //         GETARG_Bx(i)
+            //     } else {
+            //         GETARG_Ax(p.code[pc + 1])
+            //     };
+            //     if ttisstring(&p.k[b]) {
+            //         *name = Some(svalue(&p.k[b]));
+            //         return Some("constant");
+            //     }
+            // }
```
Comment (mosmeh) [inline src/runtime/debug.rs]: Unfinished?

File: src/runtime/debug.rs [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+        self.get_objname(pc, c)
+            .map(|x| if x.0 == "c" { x.1 } else { "?" })
+            .unwrap_or("??")
+    }
+
+    fn upvalname(&self, uv: usize) -> Option<&str> {
+        // TString *s = check_exp(uv < p->sizeupvalues, p->upvalues[uv].name);
+        // if (s == NULL) return "?";
+        // else return getstr(s);
+        None
```
Comment (mosmeh) [inline src/runtime/debug.rs]: Please return `?` with comments or `todo!()` here

File: src/runtime/debug.rs [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+        bottom: usize,
+    ) -> Option<(&'static str, &'a str)> {
+        // if (ci->callstatus & CIST_HOOKED) {  /* was it called inside a hook? */
+        // *name = "?";
+        // return "hook";
+        // }
+        // else if (ci->callstatus & CIST_FIN) {  /* was it called as a finalizer? */
+        // *name = "__gc";
+        // return "metamethod";  /* report it as such */
+        // }
```
Comment (mosmeh) [inline src/runtime/debug.rs]: Hooks are not handled in other parts of mochi anyway...

File: src/runtime.rs [hunk 17]
```diff
@@ -305,10 +306,18 @@ impl<'gc> Vm<'gc> {
...
                     self.push_frame(thread, bottom)
                 }
-                None => Err(ErrorKind::TypeError {
-                    operation: Operation::Call,
-                    ty: value.ty(),
-                }),
+                None => {
+                    if let Some((kind, name)) = self.funcname_from_call(thread, bottom) {
+                        Err(ErrorKind::other(format!(
+                            "attempt to call a nil value ({kind}:{name:?})"
```
Comment (mosmeh) [inline src/runtime.rs]:
```suggestion
                            "attempt to call a nil value ({kind} {name:?})"
```

**Review by mosmeh — CHANGES_REQUESTED** [review]
I had a brief look at the code, but the review may be still incomplete.

Please fix rustc and clippy warnings (you may rebase on master, as I fixed some warnings)

File: src/runtime.rs (hunk 6 above)
Comment (metaworm) [reply src/runtime.rs]: I think "last" is more semantic.

File: src/runtime/debug.rs (hunk 14 above)
Comment (metaworm) [reply src/runtime/debug.rs]: Next commit will complete it

File: src/runtime/opcode.rs (hunk 11 above)
Comment (metaworm) [reply src/runtime/opcode.rs]: Mainly to correspond with the Lua source code, I can add comments to explain the meaning

File: src/runtime/debug.rs (hunk 9 above)
Comment (metaworm) [reply src/runtime/debug.rs]: Yes, struct is more clear.

**Comment by mosmeh** [comment]
Not all points are fixed and there are clippy warnings. I will fix them by myself

**Pull request merged**